mjpc/planners/cem/planner.cc

```diff
@@ -0,0 +1,554 @@
+      });
+
+  // stop timer
+  rollouts_compute_time = GetDuration(rollouts_start);
+
+  return ncandidates;
+}
+
+// optimize nominal policy using random sampling
+void CEMPlanner::OptimizePolicy(int horizon, ThreadPool& pool) {
+  // resample nominal policy to current time
+  this->UpdateNominalPolicy(horizon);
+
+  OptimizePolicyCandidates(n_elites, horizon, pool);
+
+  // ----- update policy ----- //
+  // start timer
+  auto policy_update_start = std::chrono::steady_clock::now();
+
+  CopyCandidateToPolicy(0);
+
+  // improvement: compare nominal to winner
+  double best_return = trajectory[0].total_return;
+  improvement = mju_max(best_return - trajectory[winner].total_return, 0.0);
+
+  // stop timer
+  policy_update_compute_time = GetDuration(policy_update_start);
+}
+
+// compute trajectory using nominal policy
+void CEMPlanner::NominalTrajectory(int horizon, ThreadPool& pool) {
+  // set policy
+  auto nominal_policy = [&cp = candidate_policy[0]](
+                            double* action, const double* state, double time) {
+    cp.Action(action, state, time);
+  };
+
+  // rollout nominal policy
+  trajectory[0].Rollout(nominal_policy, task, model, data_[0].get(),
+                        state.data(), time, mocap.data(), userdata.data(),
+                        horizon);
+}
+
+// set action from policy
+void CEMPlanner::ActionFromPolicy(double* action, const double* state,
+                                       double time, bool use_previous) {
+  const std::shared_lock<std::shared_mutex> lock(mtx_);
+  if (use_previous) {
+    previous_policy.Action(action, state, time);
+  } else {
+    policy.Action(action, state, time);
+  }
+}
+
+// update policy via resampling
+void CEMPlanner::UpdateNominalPolicy(int horizon) {
+  // dimensions
+  int num_spline_points = candidate_policy[winner].num_spline_points;
+
+  // set time
+  double nominal_time = time;
+  double time_shift = mju_max(
+      (horizon - 1) * model->opt.timestep / (num_spline_points - 1), 1.0e-5);
+
+  // temp variables to help with averaging
+  std::vector<double> temp_avg(model->nu);  // holds avg control action over elites for each spline point t
+  std::vector<std::vector<double>> temp_elite_actions;
+  temp_elite_actions.resize(n_elites, std::vector<double>(model->nu, 0.0)); // shape=(n_elites, nu)
+
+  variance.assign(model->nu * num_spline_points, 0.0);  // reset variance to 0
+
+  // get spline points
+  for (int t = 0; t < num_spline_points; t++) {
+    times_scratch[t] = nominal_time;
+    temp_avg.assign(model->nu, 0.0);  // reset temp_avg to 0
+
+    // get the actions of the top n_elites policies and average them
+    // also update a variance parameter too
+    for (int i = 0; i < n_elites; i++) {
+      candidate_policy[trajectory_order[i]].Action(  // trajectory_order[i] is the (i+1)^th best trajectory
+        DataAt(temp_elite_actions[i], 0),  // copies the action from the cand policy into the i^th control vector
```

Comment:
So here's a possible segfault: if `n_elites` changes (from the GUI) between when `temp_elite_actions` was resized and here then you could ask for something out of bounds. 

My C++ multithreading is rusty, but I think the easiest fix might just be a little `const std::shared_lock<std::shared_mutex> lock(mtx_)` like you see sprinkled throughout this codebase. 

You could also/additionally consider setting the size of `temp_elite_actions` to the total possible rollouts at construction, and only use the first `n_elites` samples. But you'll still want to do the mutex lock thing to avoid looking at old/irrelevant data

Reply:
two fixes in response to this:

1. instead of using a lock, I just define a new `int n_elites_fixed = n_elites` at the top of the function that replaces all the invocations of it later on.
2. I do as you suggest and set the size of `temp_elite_actions` to `kMaxTrajectoryHorizon`